topics/data-science/tutorials/python-functions/tutorial.md

```diff
@@ -0,0 +1,300 @@
+
+
+> ### Agenda
+>
+> In this tutorial, we will cover:
+>
+> 1. TOC
+> {:toc}
+>
+{: .agenda}
+
+
+## What is a Function
+
+Functions are a way to re-use some computation you want to do, multiple times. If you don't have a function, you need to re-write the calculation every time, so we use functions to collect those statements and make them easy to re-run. Additionally they let us "parameterise" some computation. Instead of computing the same value every time, we can template it out, we can re-run the computation with new inputs, and see new results.
+
+> > ### {% icon code-in %} Math
+> > ```
+> > # Define our function
+> > f(x) = 3 * x
+> > # Compute some value
+> > f(3) # is 9
+> > ```
+> {: .code-in}
+>
+> > ### {% icon code-out %} Python
+> > ```
+> > # Define our function
+> > def f(x):
+> >    return 3 * x
+> > # Compute some value
+> > f(3)
+> > ```
+> {: .code-out}
+>
+{: .code-2col}
+
+We've talked about mathematical functions before, but now we'll talk about more programing-related functions
+
+
+## Create Functions
+
+Human beings can only keep a few items in working memory at a time. Breaking down larger/more complicated pieces of code in functions helps in understanding and using it. A function can be re-used. Write one time, use many times. (Known as staying {DRY} in the industry.)
+
+Knowing what Americans mean when they talk about temperatures and weather can be difficult, but we can wrap the temperature conversion calculation ($$^{\circ}\text{C} = (^{\circ}\text{F} - 32) * \dfrac{5}{9}$$) up as a function that we can easily re-use.
+
+```python
+def fahr_to_celsius(temp):
+    return ((temp - 32) * (5/9))
+```
+
+![The above function fahr to celsius is shown except annotated. def is labelled "def statement", fahr_to_celsius is noted as the function name. Inside parentheses is temp and an arrow shows it is called parameter names. The next line which is indented is annotated as the function body which has a return statement and the calculation from above.](../../images/python-basics/Figure7_Functions.png)
+
+The function definition opens with the keyword `def` followed by the name of the function `fahr_to_celsius` and a parenthesized list of parameter names `temp`. The body of the function — the statements that are executed when it runs — is indented below the definition line. The body concludes with a `return` keyword followed by the return value.
+
+When we call the function, the values we pass to it are assigned to those variables so that we can use them inside the function. Inside the function, we use a return statement to send a result back to whoever asked for it.
+
+Let’s try running our function.
+
+```python
+fahr_to_celsius(32)
+print('freezing point of water:', fahr_to_celsius(32), 'C')
```

Comment:
can we implement here an or just to show there are multiple options

print('freezing point of water: {0} C'.format(fahr_to_celsius(32)))

Reply:
yeahhh I wasn't sure where to put this.

- math, probably takes 1h to get everyone going?
- functions, maybe another 1h?

I don't know if there's enough time to talk about strings in the first day, or if we save that for day 2, show new, better ways to construct them.

Reply:
Changed my mind, let's do this.